src/Api/Mastodon/Account.php

```diff
@@ -80,6 +82,12 @@ public static function createFromContact(array $contact) {
 		$account->header_static = '';
 		// No custom emojis per account in Friendica
 		$account->emojis = [];
+		$account->bot = ($contact['contact-type'] == Contact::TYPE_NEWS);
+
+		$apcontact = APContact::getByURL($contact['url'], false);
```

Comment:
I'd prefer calling the `getByURL()` method before this method and pass the result (`$apcontact`) as second parameter to this method ( `createFromContact`).

So we differentiate between business logic and plain container classes.

Reply:
Then we had to do this at every location where we call it - this would be impractical.

Reply:
Does `$apcontact` have completely different fields than `$contact`?

Reply:
Some fields are identical, like the url, the nick or the name, but many are different. And we shouldn't forget: The `apcontact` only holds data from AP enabled contacts, not from Diaspora contacts. So we have to treat it as an additional data source.

Reply:
Then I'm with @nupplaphil, I think the return value of `APContact::getByURL` should be added as a parameter of `createFromContact`, optionally empty for non-AP enabled contacts.

Reply:
We don't know in advance if this is an AP contact or not. But then we had to call this function every time in the function parameters whenever we call this `createFromContact`.

Why should we make it more complicated using this function?

Reply:
Just to clarify: The improvement then is that we have to call:
``` Account::createFromContact($adminContact, APContact::getByURL($AdminContact['url'], false));``` instead of ``` Account::createFromContact($adminContact)``` on every location where we use ``createFromContact``` including the ```use Friendica\Model\APContact```?

There is never more than one apcontact entry for a contact entry. So from a database point of view this doesn't make any sense at all.

Reply:
The problem with `APContact`, as usual, is that it is a static class. This hampers tests because we can't easily mock static classes to control its inputs and outputs. A single call to `APContact::getByURL` in this method renders the method untestable.

If the code was already existing, it wouldn't be as much of a problem, because we can't refactor everything overnight to be easily testable. But for new code, we can be more rigorous regarding testability.

This is why I didn't pass a contact id, for example, although it would have been "easier" to do so and just call `Contact::getById` from the method. So yes, we could write "easier" code with less seemingly redundant lines, but sometimes it costs reliability.

Reply:
We can simply add the needed data to the database for the tests - just like we already are doing for the current API: https://github.com/friendica/friendica/blob/develop/tests/datasets/api.fixture.php

And I really don't see the difference between this and my other stuff in the API that you just accepted. There are database calls as well.

Reply:
> And I really don't see the difference between this and my other stuff in the API that you just accepted.

Because I'm choosing my battles, it would have been way more complicated to make the Instance class you created testable. However I created this Account class to be testable, and I'd like it to stay this way.

Reply:
I guess that I declare the API to my personal "hands off" area, since I don't understand the whole construct behind it.